### PR TITLE
feat: authmoji to limit the actions an api token can perfrom

### PR DESCRIPTION
**Pull request by olizilla**
new authmoji! it's cool! It's almost exactly what was asked for in #231

```js
const auth = {
  '🤲': handler => withCorsHeaders(handler),
  '🔒': handler => withCorsHeaders(withApiOrMagicToken(handler)),
  '👮': handler => withCorsHeaders(withMagicToken(handler))
}

router.post('/user/login',          auth['🤲'](userLoginPost))
router.get('/status/:cid',          auth['🤲'](statusGet))
router.get('/car/:cid',             auth['🤲'](carGet))
router.head('/car/:cid',            auth['🤲'](carHead))

router.post('/car',                 auth['🔒'](carPost))
router.put('/car/:cid',             auth['🔒'](carPut))
router.post('/upload',              auth['🔒'](uploadPost))
router.get('/user/uploads',         auth['🔒'](userUploadsGet))

router.delete('/user/uploads/:cid', auth['👮'](userUploadsDelete))
router.get('/user/tokens',          auth['👮'](userTokensGet))
router.post('/user/tokens',         auth['👮'](userTokensPost))
router.delete('/user/tokens/:id',   auth['👮'](userTokensDelete))
router.get('/user/account',         auth['👮'](userAccountGet))
```

less cool is that we need some mechanism to bypass magic.link in testing so this PR adds a `DANGEROUSLY_BYPASS_MAGIC_AUTH` env var that is only set in the worker-globals.js test code, and no where near prod. It's not great, but it it would be difficult to set accidentally.

License: (Apache-2.0 AND MIT)
Signed-off-by: Oli Evans <oli@tableflip.io>

### CODE REVIEW

File: packages/api/src/auth.js [hunk 1]
```diff
@@ -0,0 +1,121 @@
...
+ */
+export function withMagicToken (handler) {
+  /**
+   * @param {Request} request
+   * @param {import('./env').Env}
+   * @returns {Response}
+   */
+  return async (request, env, ctx) => {
+    const token = getTokenFromRequest(request, env)
+    if (isMagicToken(token, env)) {
```
Comment (alanshaw) [inline packages/api/src/auth.js:26]:
```suggestion
    if (await isMagicToken(token, env)) {
```

File: packages/api/src/auth.js [hunk 2]
```diff
@@ -0,0 +1,121 @@
...
+   * @param {Request} request
+   * @param {import('./env').Env}
+   * @returns {Response}
+   */
+  return async (request, env, ctx) => {
+    const token = getTokenFromRequest(request, env)
+    if (!token) {
+      throw new NoTokenError()
+    }
+    if (isWeb3ApiToken(token, env)) {
```
Comment (alanshaw) [inline packages/api/src/auth.js:57]:
```suggestion
    if (await isWeb3ApiToken(token, env)) {
```

File: packages/api/src/auth.js [hunk 3]
```diff
@@ -0,0 +1,121 @@
...
+      const authToken = await findWeb3ApiToken(token, env)
+      if (!authToken) {
+        throw new TokenNotFoundError()
+      }
+      request.auth = { authToken, user: authToken.user }
+      env.sentry && env.sentry.setUser(authToken.user)
+      return handler(request, env, ctx)
+    }
+    if (isMagicToken(token, env)) {
```
Comment (alanshaw) [inline packages/api/src/auth.js:66]:
```suggestion
    if (await isMagicToken(token, env)) {
```

**Review by alanshaw — COMMENTED** [review]
Oops, no review requested, stopping! Leaving feedback so far just incase it's useful.

File: packages/api/src/auth.js (hunk 1 above)
Comment (olizilla) [reply packages/api/src/auth.js:26]: almost! turns out `magic.token.validate` is sync, and will throw if it doesn't recognise it. I'm going to catch that error and return false.

**Comment by olizilla** [comment]
oh no! 
> Coming soon! Admin-sdk support for testMode!
– https://magic.link/docs/introduction/test-mode

**Comment by olizilla** [comment]
@alanshaw please to 👀  now it has tests! that pass!

**Review by alanshaw — APPROVED** [review]
LGTM

**Pull request merged**